[Asset] Adding a new version strategy that reads from a manifest JSON file

weaverryan opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | n/a
| License       | MIT
| Doc PR        | https://github.com/symfony/symfony-docs/issues/7659

Hi guys!

Often, when using a frontend task manager or bundler (e.g. webpack of gulp), the final assets are dumped with a version or content hash in the filename itself (e.g. main.123abc.css). To know what the correct, current hashed filename is, you'll dump a `manifest.json` file - e.g.

```json
{
  "main.js": "main.123abc.js",
  "css/styles.css": "css/styles.555def.css"
}
```

Examples: [gulp-rev](https://github.com/sindresorhus/gulp-rev) and [webpack-manifest-plugin](https://www.npmjs.com/package/webpack-manifest-plugin).

This PR adds a new version strategy that will look up the asset path (e.g. `main.css`) in that file and return the final, versioned path. Some people may dump manifest files in other formats, but I think this catches the most common use-case (and you can always still create your own version strategy). I've written this to be "forgiving" - if a path doesn't exist in the manifest, the path is simply returned, unaltered.

Another implementation *could* have been to add a new Twig filter (e.g. `{{ asset('main.css|manifest_path) }}`) - but I thought I'd try first using the existing versioning system.

## Usage

```yml
# app/config/config.yml
framework:
    # ...
    assets:
       # added validation prevents you from setting json_manifest_path AND version, for example
        json_manifest_path: '%kernel.root_dir%/../web/manifest.json'
```

```twig
{# someTemplate.html.twig #}

{# use asset() just like normal #}
<script src="{{ asset('js/main.js') }}"></script>
```

## TODO
* fabbot hates my invalid json syntax file... even though I tried to be clever and not give it a `.json` suffix :)

weaverryan committed:
[Asset][FrameworkBundle] Adding a new version strategy that reads from a JSON manifest file

weaverryan committed:
fabbot

weaverryan committed:
Removing .json from extension of a file to work around fabbot not liking the invalid JSON

weaverryan committed:
Fixing a test

javiereguiluz commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php:
```diff
@@ -589,6 +597,12 @@ private function addAssetsSection(ArrayNodeDefinition $rootNode)
                                 ->end()
+                                ->validate()
+                                    ->ifTrue(function ($v) {
+                                        return isset($v['version_strategy']) && $v['version_strategy'] == 'json_manifest' && !isset($v['manifest_path']);
```
`== 'json_manifest'` -> `=== 'json_manifest'`

javiereguiluz commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php:
```diff
@@ -558,6 +559,12 @@ private function addAssetsSection(ArrayNodeDefinition $rootNode)
                     ->end()
+                    ->validate()
+                        ->ifTrue(function ($v) {
+                            return isset($v['version_strategy']) && $v['version_strategy'] == 'json_manifest' && !isset($v['manifest_path']);
```
`== 'json_manifest'` -> `=== 'json_manifest'`

javiereguiluz reviewed (approved):
A big 👍  from me!

This is so common in modern setups that Symfony should really provide this asset versioning.

I also agree with being forgiving and returning the same path when it doesn't exist in the manifest file.

weaverryan committed:
== => ===

weaverryan committed:
Another attempt at faking out fabbot. Quick fabbot - look over there!

(fabbot dislikes invaild JSON)

javiereguiluz commented on src/Symfony/Component/Asset/Tests/fixtures/manifest-valid.json:
```diff
@@ -0,0 +1,4 @@
+{
+  "main.js": "main.123abc.js",
+  "css/styles.css": "css/styles.555def.css"
+}
```
We should add a new line at the end of this file.

javiereguiluz commented on src/Symfony/Component/Asset/Tests/fixtures/manifest-invalid-json:
```diff
@@ -0,0 +1,4 @@
+{
+  "main.js": main.123abc.js",
+  "css/styles.css": "css/styles.555def.css"
+}
```
We should add a new line at the end of this file.

javiereguiluz commented on src/Symfony/Component/Asset/Tests/fixtures/manifest-invalid-json:
```diff
@@ -0,0 +1,4 @@
+{
```
It's clever to not use `.json` extension to avoid fabbot issues ... but that could complicate the maintenance of this. Imagine looking at this conde in 1 year and wondering: why this file doesn't have an extension? Is this an error? Should I add the extension now? etc.

So, let's add the `.json` extension and forget about fabbot issues.

thewilkybarkid commented:
Looks to replace https://github.com/irozgar/gulp-rev-versions-bundle.

A useful feature would be to compile the manifest into a static array, to avoid reading it at runtime.

stof commented:
This changes the meaning of the ``version_strategy`` configuration, as this is not a service id anymore in some cases.
IMO, the usage of the json manifest strategy should be detected based only on `manifest_path` being configured and no custom strategy being provided, similar to the way the static version strategy is configured

stof commented on src/Symfony/Component/Asset/VersionStrategy/JsonManifestVersionStrategy.php:
```diff
@@ -0,0 +1,72 @@
+    {
+        $manifestPath = $this->getManifestPath($path);
+
+        return $manifestPath ? $manifestPath : $path;
```
``return $manifestPath ?: $path;``

stof commented on src/Symfony/Component/Asset/VersionStrategy/JsonManifestVersionStrategy.php:
```diff
@@ -0,0 +1,72 @@
+
+    private $manifestData;
+
+    public function __construct($manifestPath)
```
please add the phpdoc describing the argument as a string, as we don't have a typehint to tell it

stof commented on src/Symfony/Component/Asset/VersionStrategy/JsonManifestVersionStrategy.php:
```diff
@@ -0,0 +1,72 @@
+            }
+
+            $this->manifestData = json_decode(file_get_contents($this->manifestPath), true);
+            if (0 < $errorCode = json_last_error()) {
```
`$errorCode` is unused

stof commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -857,6 +865,17 @@ private function createVersion(ContainerBuilder $container, $version, $format, $
+    {
+        $def = new ChildDefinition('assets.json_manifest_version_strategy');
+        $def
+            ->replaceArgument(0, $manifestPath)
```
IMO, we should also make this service private, to allow it to be inlined

weaverryan replied:
I agree, but I copied this code from the above method, which is not private. So, might there be a reason for it to be public? https://github.com/symfony/symfony/blob/4927993835457bc2ae23826f5572f97e26dda0ec/src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php#L843-L856

I'm going to try making it private.

weaverryan committed:
Line breaks and not being clever with invalid - just ending with .json

weaverryan committed:
Tweaks thanks to Stof!

GromNaN commented:
Awesome feature that I've seen implemented in many projects, this JSON format is so simple that it is always the one that is used.
Is it possible to have many manifests ? I remember having JS, CSS and images assets in different manifests files. But also mobile and desktop assets in different manifests.

Is the solution to define an asset package for each manifest file ?

javiereguiluz commented:
@GromNaN is having multiple manifest still common? I only know the Webpack solution and they create just 1 `manifest.json` for all assets(images, CSS, JS).

weaverryan committed:
Reworking feature so that json_manifest_path is the config key, and is all that's needed to activate

This keeps the original behavior of version_strategy: this is always a service id

weaverryan committed:
Thanks fabbot!

weaverryan committed:
Making test compatible with older phpunit versions

weaverryan commented:
PR Updated!

1) I like Stof's [suggestion](https://github.com/symfony/symfony/pull/22046#issuecomment-288030843) about keeping `version_strategy` always as an `id`. I updated the config usage in the PR description

2)  @thewilkybarkid I don't know if compiling a static array from the JSON is a good idea or not. It might be unexpected that simply updating the file wouldn't cause Symfony to pick up the changes. I'm not sure if the decoding overhead is significant enough to care - decoding a 30 line manifest takes `.00011` seconds (`.11` milliseconds), though it is a filesystem hit!

3) @GromNaN That's interesting, I want to make sure we've at least thought through this scenario (though ultimately, the strategy doesn't need to work for *everyone*. In general, can you think of a solution *other* than multiple packages? Even if we made this feature much bigger and supported multiple manifest files, in Twig, you still would need to specify which one you need, and the only config you can pass to `asset()` is the `$packageName`. Or, the strategy would need to somehow be smart enough to *know* which manifest file to use for each package, which now sounds like you would/should have a custom version strategy. 

This is ready to be reviewed again! Thanks!

javiereguiluz commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php:
```diff
@@ -582,6 +595,7 @@ private function addAssetsSection(ArrayNodeDefinition $rootNode)
                                         ->end()
                                     ->end()
                                     ->scalarNode('version_format')->defaultNull()->end()
+                                    ->scalarNode('json_manifest_path')->defaultNull()->end()
```
Just asking: would `manifest_path` be a more future-proof name than `json_manifest_path` ? Or is this too much future thinking? Can we expect to support other formats different than JSON?

weaverryan replied:
It's a good thought. I don't know! It's a fairly easy thing to deprecate and change in the future anyways, and I can't see any other format in the near future. So let's go with what we like best right now. I changed to `json_manifest_path` because I thought it might be more clear what this manifest thing should be.

nicolas-grekas replied:
lgtm as json*

weaverryan committed:
Fixing test - the old way is all inside setExpectedException, the new way is in 2 method calls

GromNaN commented:
@javiereguiluz @weaverryan My actual use case is the following:
Having a website on www.website.com with an mirror mobile site m.website.com 
Routing and controllers are the same for both domains, but depending on the domain the view listener render a different template (thanks to Twig namespaces).
Both websites are very different in terms of design, that why the frontend developers chose to have 2 separate builds, generating 2 separate JSON manifests.

---

Regarding the usage of a PHP file to cache the content of the JSON into a PHP array, from my experience on high traffic websites there is no performance issue at reading the same JSON file on each request: the OS caches the file content in memory and native JSON parsing is very fast.

weaverryan commented:
@GromNaN Hmm, in that case, it would be most convenient for the listener to be able to change the manifest file. If you created a default package and a `mobile` package, then in the listener, if you're on mobile, you could make the `mobile` package the default:

```php
// in the listener... if mobile
// using the container for clarity
$packages = $this->container->get('assets.packages');
$packages->setDefaultPackage($packages->getPackage('mobile'));
```

And then your templates wouldn't need to specify a package everywhere. Of course, since they're using different templates, they could just specify the `mobile` package when using `asset()` instead of the above code.

I think this is actually a pretty decent solution. In other words, I think this PR solves the 99%, and makes the 1% reasonably easy.

Thanks for the reply :)

fabpot commented:
@weaverryan For the invalid JSON file, just add it to `.php_cs.dist`

fabpot commented on src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md:
```diff
@@ -4,6 +4,8 @@ CHANGELOG
 3.3.0
 -----
 
+ * A new version strategy option called json_manifest_path was added
```
added a new ...

fabpot commented on src/Symfony/Component/Asset/CHANGELOG.md:
```diff
@@ -1,6 +1,11 @@
 
+3.3.0
+-----
+ * JsonManifestVersionStrategy was added as a way to read final,
```
added Json...

fabpot commented on src/Symfony/Component/Asset/VersionStrategy/JsonManifestVersionStrategy.php:
```diff
@@ -0,0 +1,75 @@
+ *         "css/styles.css": "css/styles.555abc.css"
+ *     }
+ *
+ * You could then as for the version of "main.js" or "css/styles.css".
```
typo: ask?

fabpot commented on src/Symfony/Component/Asset/VersionStrategy/JsonManifestVersionStrategy.php:
```diff
@@ -0,0 +1,75 @@
+class JsonManifestVersionStrategy implements VersionStrategyInterface
+{
+    private $manifestPath;
+
```
empty line can be removed

fabpot commented on src/Symfony/Component/Asset/VersionStrategy/JsonManifestVersionStrategy.php:
```diff
@@ -0,0 +1,75 @@
+    private $manifestData;
+
+    /**
+     * @param string $manifestPath Absolute path to the manifest file.
```
no dot at the end

fabpot commented on src/Symfony/Component/Asset/VersionStrategy/JsonManifestVersionStrategy.php:
```diff
@@ -0,0 +1,75 @@
+     *
+     * @param string $path
+     *
+     * @return string
```
both @param and @return can be removed

GromNaN commented:
The manifest could be a remote file accessible over HTTP ; that appends when assets and the app are deployed on distinct servers asynchronously. In that case, it is necessary to use an HTTP client (other that file_get_contents) and a cache layer.

fabpot commented:
@GromNaN I would say that we don't want to support that use case.

weaverryan committed:
Tweaks thanks to fabpot

fabpot commented:
👍

weaverryan commented:
@GromNaN I thought about that use-case, it's also relevant with webpack-dev-server, but ultimately, I didn't want it, at least for now: the user should make sure there's a physical JSON file - it could be part of their deploy process (or, they can of course add their own strategy!)

Tweaks made! But fabbot and I are still fighting - is the JSON check done through `PhpCsFixer`? Or did I just not get the config right in `.php_cs.dist`?

weaverryan commented:
@fabpot if my silly invalid JSON file is going to be a problem with fabbot, let me know - I could re-work things - i.e. save a tmp invalid JSON file in the test and use that (so no committed invalid JSON file).

fabpot commented:
@weaverryan Don't worry about that. I will see why it's not taken into account by fabbot. Keep the code as is.

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md:
```diff
@@ -4,6 +4,8 @@ CHANGELOG
 3.3.0
 -----
 
+ * Added a new new version strategy option called json_manifest_path
```
quotes around "json_manifest_path"?

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md:
```diff
@@ -4,6 +4,8 @@ CHANGELOG
 -----
 
+ * Added a new new version strategy option called json_manifest_path
+   that allows you to use the JsonManifestVersionStrategy.
```
backticks here?

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -856,20 +859,30 @@ private function createPackageDefinition($basePath, array $baseUrls, Reference $
+        if (null !== $jsonManifestPath) {
+            $def = new ChildDefinition('assets.json_manifest_version_strategy');
+            $def->replaceArgument(0, $jsonManifestPath);
+            $def->setPublic(false);
```
move to xml?

nicolas-grekas commented on src/Symfony/Component/Asset/VersionStrategy/JsonManifestVersionStrategy.php:
```diff
@@ -0,0 +1,70 @@
+    {
+        $manifestPath = $this->getManifestPath($path);
+
+        return $manifestPath ?: $path;
```
could be a oneliner :)

nicolas-grekas reviewed (approved):
:+1: with optional comments

weaverryan committed:
Very minor changes thanks to nicolas-grekas

weaverryan commented:
Minor comments addressed!

fabpot commented:
Thank you @weaverryan.
